src/solver/README.md

```diff
@@ -0,0 +1,238 @@
+
+### Backtracking
+This optimization allows to benefit from the recognition of the fact that most resources are only available in some sources.
+Instead of generating all possible sets and then ignoring ones which are incomplete, it allows the algorithm to [backtrack](https://en.wikipedia.org/wiki/Backtracking) from partial candidates that cannot lead to a complete solution.
+
+That technique is very powerful in the `L10nRegistry` use case and in many scenarios leads to 10-100x speed ups even in cases where all sets have to be generated.
+
+# Serial vs Parallel Testing
+At the core of the solver is a *tester* component which is responsible for eagerly evaluating candidates to allow for early bailouts from partial solutions which cannot lead to a complete solution.
+
+This can be performed in one of two ways:
+
+### Serial
+ The algorithm is synchronous and each extension of the candidate is evaluated serially, one by one, allowing the for *backtracking* as soon as a given extension of a partial solution is confirmed to not lead to a complete solution.
+
+Bringing back the initial state of the solver:
+ 
+```text
+  ▼
+┌┲━┱┬───┬───┐
+│┃0┃│   │   │
+└╂─╂┴───┴───┘
+ ┃ ┃     
+ ┗━┛
+```
+
+The tester will evaluate whether the first resource **A** is available in the first source **0**. The testing will be performed synchronously, and the result will inform the algorithm on whether the candidate may lead to a complete solution, or this branch should be bailed out from, and the next candidate must be tried.
+
+#### Success case
+
+If the test returns a success, the extensions of the candidate is generated:
+```text
+      ▼
+┌┲━┱┬┲━┱┬───┐
+│┃0┃│┃0┃│   │
+└╂─╂┴╂─╂┴───┘
+ ┃ ┃ ┃ ┃ 
+ ┗━┛ ┗━┛
+```
+
+When a candidate is complete, in other words, when the last cell of a candidate has been tested and did not lead to a backtrack, we know that the candidate is a solution to the problem, and we can yield it from the iterator.
+
+#### Failure case
+
+If the test returns a failure, the next step is to evaluate alternative source for the same resource. Let's assume that *Source 0* had *Resource A* but it does not have *Resource B*. In such case, the algorithm will increment the cell's index:
+
+```text
+      ▼
+     ┏━┓
+     ┃0┃
+┌┲━┱┬╂─╂┬───┐
+│┃0┃│┃1┃│   │
+└╂─╂┴┺━┹┴───┘
+ ┃ ┃    
+ ┗━┛ 
+ ```
+
+and that will potentially lead to a partial solution `[0, 1, ]` to be stored for the next iteration.
+
+If the test fails and no more sources can be generated, the algorithm will *backtrack* from to previous sources looking for a cell with the **highest** index prior to the cell that was being evaluated which is not yet on the last source. If such cell is found, the results of all cells **to the right** of the newfound cell are **erased** and the next branch can be evaluated.
```

Comment:
Does highest index here mean the source where you have the longest known potential solution?

nit: There is a missing word after *backtrack* from... I assume *backtrack* from the current cell.

Reply:
No, it means that if we are on a cell idx 3, we want to find the highest cell with index lower than 3 that is not yet "maxed out".

So, we'll see if cell idx 2 can have source bumped. If not, if cell idx 1 can etc. The first lower than 3 that we find is where we stop backtracking, because now we can bump it's source and move forward again.

How would you phrase such sentence? :)

Reply:
Maybe I understand better now: so if it fails, we backtrack from right to left on the partial solution, looking for a cell that is not "maxed out". If we find such a cell, we increment it, and "prune" everything to the right of it from the partial solution?

So [0, 0, 1, 1] would back track to [0, 0, ], increment to [0, 1], and get rid of everything to the right?

I was tripping up over cell index vs the cell contents and whether highest was the cell index or referred to a cell whose contents were the highest source.

Reply:
Yeah! Maybe we should add it to the legend at the top?
I was also asked to consider generalizing it and releasing as a standalone and maybe white paper.
If so, we'd want to move away from "resources" to "keys" or some other term.
I think we can keep "source" for the other axis and "test" for the memoized operation.

Reply:
That make sense. I think using a word other than "resources" would help with readability, I found it difficult to keep track of which was which with "resource" and "source" just because of the similarity in the words.

Reply:
I'll change that when I'll be looking to generalize the algo and extract it from this crate. For now I think it's ok to keep resources/sources.